rust/common/cookieless/src/manager.rs

```diff
@@ -0,0 +1,895 @@
+                // If there's a Redis error, propagate it
+                Err(CookielessManagerError::RedisError(e.to_string()))
+            }
+        }
+    }
+}
+
+/// Extract the root domain from a host string
+///
+/// This function handles various formats including:
+/// - URLs with protocols (e.g., https://example.com)
+/// - Hosts with ports (e.g., example.com:8000)
+/// - Subdomains (e.g., sub.example.com)
+/// - IPv4 and IPv6 addresses
+///
+/// It returns the root domain (eTLD+1) for valid domains, or the original host for
+/// special cases like IP addresses, localhost, etc.
+/// The port is preserved if present in the original host.
+pub fn extract_root_domain(host: &str) -> Result<String, CookielessManagerError> {
+    use std::net::IpAddr;
+
+    // If the host is empty, return it as is
+    if host.is_empty() {
+        return Ok(host.to_string());
+    }
+
+    // Check if it's an IPv6 address
+    if let Ok(IpAddr::V6(ipv6)) = host.parse::<IpAddr>() {
+        // Return the normalized form of the IPv6 address in brackets
+        return Ok(format!("[{}]", ipv6));
+    }
+
+    // Split host and port for special handling of IP addresses
+    let (hostname, port) = if let Some((h, p)) = host.split_once(':') {
+        (h, Some(p))
+    } else {
+        (host, None)
+    };
+
+    // Check if the hostname is an IP address
+    if hostname.parse::<IpAddr>().is_ok() {
+        return match port {
+            Some(p) => Ok(format!("{}:{}", hostname, p)),
+            None => Ok(hostname.to_string()),
+        };
+    }
+
+    // Add a fake protocol if none exists
+    let input = if !host.contains("://") {
+        format!("http://{}", host)
+    } else {
+        host.to_string()
+    };
+
+    // Parse the URL to extract hostname and port
+    let url = match Url::parse(&input) {
+        Ok(url) => url,
+        Err(_) => return Ok(host.to_string()),
+    };
+
+    let hostname = url.host_str().unwrap_or(host).to_string();
+    let port = url.port().map(|p| p.to_string());
+
+    // Use the public-suffix list to extract the root domain (eTLD+1)
+    let domain = match DEFAULT_PROVIDER.effective_tld_plus_one(&hostname) {
+        Ok(domain) => domain.to_string(),
+        Err(_) => hostname,
+    };
+
+    // Add the port back if it exists
+    match port {
+        Some(p) => Ok(format!("{}:{}", domain, p)),
+        None => Ok(domain),
+    }
+}
+
+/// Convert a timestamp to YYYY-MM-DD format in the specified timezone
+fn to_yyyy_mm_dd_in_timezone_safe(
+    timestamp_ms: u64,
+    event_time_zone: Option<&str>,
+    team_time_zone: &str,
+) -> Result<String, CookielessManagerError> {
+    // Use the event timezone if provided, otherwise fall back to the team timezone
+    let timezone_str = event_time_zone.unwrap_or(team_time_zone);
+
+    // Parse the timezone
+    let timezone = match chrono_tz::Tz::from_str(timezone_str) {
+        Ok(tz) => tz,
+        Err(_) => match chrono_tz::Tz::from_str(TIMEZONE_FALLBACK) {
```

Comment:
Can't we have this as a comptime constant rather than having to compute it every time? I'm pretty sure Rust will let us do that

Reply:
That'll let us remove the `TimezoneError`

Reply:
K this is way better now